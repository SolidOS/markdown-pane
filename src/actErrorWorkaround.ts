/* eslint-env jest */

/* istanbul ignore next [This is a test helper, so it doesn't need to be tested itself] */
/**
 * This is a workaround for a bug that will be fixed in react-dom@16.9
 *
 * The bug results in a warning being thrown about calls not being wrapped in `act()`
 * when a component calls `setState` twice.
 * More info about the issue: https://github.com/testing-library/react-testing-library/issues/281#issuecomment-480349256
 * The PR that will fix it: https://github.com/facebook/react/pull/14853
 */
export function workaroundActError () {
  const originalError = console.error
  beforeAll(() => {
    console.error = (...args) => {
      if (/Warning.*not wrapped in act/.test(args[0])) {
        return
      }
      originalError.call(console, ...args)
    }
  })

  afterAll(() => {
    console.error = originalError
  })
}
