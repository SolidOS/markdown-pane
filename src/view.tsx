import * as React from "react";
import Markdown from "react-markdown";

interface Props {
  markdown: string;
  onSave: (newMarkdown: string) => Promise<Response>;
}

export const View: React.FC<Props> = (props) => {
  const [phase, setPhase] = React.useState<"saving" | "rendering" | "editing">(
    "rendering"
  );
  const [rawText, setRawText] = React.useState(props.markdown);

  function storeMarkdown() {
    setPhase("saving");
    props.onSave(rawText).then(() => {
      setPhase("rendering");
    });
  }

  if (phase === "saving") {
    return <span aria-busy={true}>Loading&hellip;</span>;
  }

  if (phase === "editing") {
    return (
      <form
        onSubmit={(e) => {
          e.preventDefault();
          storeMarkdown();
        }}
      >
        <textarea
          onChange={(e) => {
            setRawText(e.target.value);
          }}
          defaultValue={rawText}
        />
        <button type="submit">RENDER</button>,
      </form>
    );
  }

  return (
    <form
      onSubmit={(event) => {
        event.preventDefault();
        setPhase("editing");
      }}
    >
      <Markdown source={rawText} />
      <button type="submit">EDIT</button>
    </form>
  );
};
