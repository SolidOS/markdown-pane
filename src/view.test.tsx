/* eslint-env jest */
import * as React from 'react'
import {
  render,
  fireEvent
} from '@testing-library/react'
import { View } from './view'
import { workaroundActError } from './actErrorWorkaround'

workaroundActError()

it('should properly render markdown', () => {
  const { container } = render(<View markdown='Some **awesome** markdown' onSave={jest.fn()}/>)

  expect(container).toMatchSnapshot()
})

describe('Edit mode', () => {
  it('should properly render the edit form', () => {
    const { container, getByRole } = render(<View markdown='Arbitrary markdown' onSave={jest.fn()}/>)

    const editButton = getByRole('button')
    editButton.click()

    expect(container).toMatchSnapshot()
  })

  it('should call the onSave handler after saving the new content', () => {
    const mockHandler = jest.fn().mockReturnValue(Promise.resolve())
    const { getByRole, getByDisplayValue } = render(<View markdown='Arbitrary markdown' onSave={mockHandler}/>)

    const editButton = getByRole('button')
    editButton.click()

    const textarea = getByDisplayValue('Arbitrary markdown')
    fireEvent.change(textarea, { target: { value: 'Some _other_ markdown' } })

    const renderButton = getByRole('button')
    renderButton.click()

    expect(mockHandler.mock.calls.length).toBe(1)
    expect(mockHandler.mock.calls[0][0]).toBe('Some _other_ markdown')
  })
})
