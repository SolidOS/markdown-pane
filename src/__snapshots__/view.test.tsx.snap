// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit mode should properly render the edit form 1`] = `
<div>
  <form>
    <textarea>
      Arbitrary markdown
    </textarea>
    <button
      type="submit"
    >
      RENDER
    </button>
    ,
  </form>
</div>
`;

exports[`should properly render markdown 1`] = `
<div>
  <form>
    <p>
      Some 
      <strong>
        awesome
      </strong>
       markdown
    </p>
    <button
      type="submit"
    >
      EDIT
    </button>
  </form>
</div>
`;
